Adding Container Metrics for gobblin-yarn

sahilTakiar opened the pull request:
* YARN containers were not reporting any Task Level Metrics (e.g. records extracted, records written, bytes written, etc.) because Metrics Reporting was not enabled in the containers
* Added `ContainerMetrics.java` to handle metrics reporting for each YARN container
* The metrics hierarchy is as follows `ContainerMetrics` -> `JobMetrics` -> `TaskMetrics`
* A `JobMetrics` instance is obtained in each `GobblinHelixTask`
* Unlike `MRJobLauncher` where the `JobMetrics` instance is manually removed from the `GobblinMetricsRegistry`, `GobblinHelixTask` does not remove the `JobMetrics` instance at the end of each task because there may be more tasks in the container that need the `JobMetrics` instance
* Reporting is done on the `ContainerMetrics` instance
* Testing is still in progress

@liyinan926 and @ibuenros can you review?

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/util/JobMetrics.java:
```diff
@@ -35,11 +34,15 @@
-    super(name(job), null, tagsForJob(job));
-    this.jobName = job.getJobName();
+    this(job, null);
 
```
Additional empty line.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/util/JobMetrics.java:
```diff
@@ -62,6 +65,17 @@ public static JobMetrics get(String jobId) {
   /**
+   * Get a new {@link GobblinMetrics} instance for a given job.
+   *
+   * @param jobState he given {@link JobState} instance
```
Typo `he` => `the`.

liyinan926 commented on gobblin-yarn/src/main/java/gobblin/yarn/ContainerMetrics.java:
```diff
@@ -0,0 +1,55 @@
+package gobblin.yarn;
```
Missing open source header.

liyinan926 commented on gobblin-yarn/src/main/java/gobblin/yarn/ContainerMetrics.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.util.List;
+
+import com.google.common.collect.Lists;
```
Wrong order of imports.

sahilTakiar replied:
What's the right order?

liyinan926 replied:
`org` should be above `com`. https://github.com/linkedin/gobblin/wiki/CodingStyle.

liyinan926 commented on gobblin-yarn/src/main/java/gobblin/yarn/ContainerMetrics.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  private static List<Tag<?>> tagsForContainer(State containerState, String applicationName, ContainerId containerId) {
+    List<Tag<?>> tags = Lists.newArrayList();
```
You can use a `ImmutableList`.

liyinan926 commented on gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTaskFactory.java:
```diff
@@ -14,16 +14,21 @@
 import java.io.IOException;
 
+import com.codahale.metrics.Counter;
+
```
Wrong order of imports.

liyinan926 commented on gobblin-yarn/src/main/java/gobblin/yarn/GobblinWorkUnitRunner.java:
```diff
@@ -305,11 +332,13 @@ public ParticipantShutdownMessageHandler(Message message, NotificationContext co
-            messageSubType.equalsIgnoreCase(HelixMessageSubTypes.WORK_UNIT_RUNNER_SHUTDOWN.toString()),
-            String.format("Unknown %s message subtype: %s", Message.MessageType.SHUTDOWN.toString(), messageSubType));
+        Preconditions
+            .checkArgument(messageSubType.equalsIgnoreCase(HelixMessageSubTypes.WORK_UNIT_RUNNER_SHUTDOWN.toString()),
```
The way the lines are broken looks weird.

sahilTakiar replied:
I agree but according to Intellij it follows the Gobblin coding style.

liyinan926 commented:
Mostly formatting comments, otherwise LGTM.

liyinan926 commented:
Seems the build was hanging.

sahilTakiar commented:
Yes, it hangs locally too. Looking into it still.

sahilTakiar commented:
I'm still addressing the comments and fixing the build. However, I tested this on Nertz and it does not seem to be working. No metrics are being emitted and I am not sure why. It only works if I invoke `jobMetrics.startMetricsReporting`, but not when I do `containerMetrics.startMetricsReporting`.

@liyinan926, @ibuenros do either of you see anything wrong with this implementation?

My understanding is that since `ContainerMetrics` is the parent of `JobMetrics` and, `JobMetrics` is the parent of `TaskMetrics`, that if I invoke `containerMetrics.startMetricsReporting` everything should be reported properly.

sahilTakiar commented:
@liyinan926 and @ibuenros I found the issue.

It seems that in Typesafe's Config library the `.conf` files actually allow you to express the config in JSON format (according to the documentation on their Github page it actually uses a JSON superset called HOCON https://github.com/typesafehub/config#using-hocon-the-json-superset).

HOCON provides multiple ways of writing valid JSON (check the documentation link for examples). One such way is using the dot-notation. So a config such as `my.test.config=30` will actually be converted to

```
{
  "my" : {
    "test" : {
      "config" : "30"
    }
  }
}
```

The reason that this is problematic is that if you have the following two lines in your config file:

```
metrics.reporting.kafka.topic=MetricReport
metrics.reporting.kafka.topic.events=GobblinTrackingEvent
```

Then the resulting config is:

```
{
  "metrics": {
    "reporting": {
      "kafka": {
        "topic": {
          "events": "GobblinTrackingEvent"
        }
      }
    }
  }
}
```

My understanding is that you the first config sets a node in the JSON tree to `MetricReport`, but then the second config over-writes that node to `{"events" : "GobblinTrackingEvent"}`.

This is exactly the bug that I was hitting when testing this on Nertz. No metrics were being reported because the code could not find the name of the Kafka topic to report to. What happens is that `GobblinMetrics.startMetricsReporting` takes in a `Properties` object. So we have to use `YarnHelixUtils.configToProperties` to convert a `Config` object to a `Properties` object. The method uses `Config.entrySet` to get all the configs in the `Config` object. It seems as though `entrySet` will only return the leaf values of the JSON tree, so the resulting `Properties` object has nothing set for `metrics.reporting.kafka.topic`.

Is this behavior expected? Do we know of a way around this? This seems very different from how `gobblin.configuration.State` works, so it seems we could easily hit this issue with other config keys?

For now I can unblock this PR, because it seems the `gobblin-metrics` uses 3 config keys for reporting data to Kafka `metrics.reporting.kafka.topic.metrics`, `metrics.reporting.kafka.topic.events`, and `metrics.reporting.kafka.topic`. The last value is used as the default if either of the other two are not set. So we can still get things to work, but I still think we should discuss this on Monday.

sahilTakiar commented:
The issue above has been discussed offline.

PR updated to address all comments. Confirmed this is working as expected after testing on Nertz.

@liyinan926, @ibuenros any more comments?

liyinan926 commented:
LGTM.
